src/frontend/src/optimizer/plan_node/stream_source_backfill.rs

```diff
@@ -0,0 +1,186 @@
+            Distribution::SomeShard,
+            source.catalog.as_ref().map_or(true, |s| s.append_only),
+            false,
+            FixedBitSet::with_capacity(source.column_catalog.len()),
+        );
+
+        Self { base, core: source }
+    }
+
+    fn get_columns(&self) -> Vec<&str> {
+        self.core
+            .column_catalog
+            .iter()
+            .map(|column| column.name())
+            .collect()
+    }
+
+    pub fn source_catalog(&self) -> Rc<SourceCatalog> {
+        self.core
+            .catalog
+            .clone()
+            .expect("source backfill should have source cataglog")
+    }
+
+    pub fn infer_internal_table_catalog() -> TableCatalog {
+        // note that source's internal table is to store partition_id -> offset mapping and its
+        // schema is irrelevant to input schema
+        // On the premise of ensuring that the materialized_source data can be cleaned up, keep the
+        // state in source.
+        // Source state doesn't maintain retention_seconds, internal_table_subset function only
+        // returns retention_seconds so default is used here
+        let mut builder = TableCatalogBuilder::default();
+
+        let key = Field {
+            data_type: DataType::Varchar,
+            name: "partition_id".to_string(),
+            sub_fields: vec![],
+            type_name: "".to_string(),
+        };
+        let value = Field {
+            data_type: DataType::Jsonb,
+            name: "backfill_progress".to_string(),
+            sub_fields: vec![],
+            type_name: "".to_string(),
+        };
+
+        let ordered_col_idx = builder.add_column(&key);
+        builder.add_column(&value);
+        builder.add_order_column(ordered_col_idx, OrderType::ascending());
+        // read prefix hint is 0. We need to scan all data in the state table.
+        builder.build(vec![], 0)
+    }
+
+    pub fn adhoc_to_stream_prost(
+        &self,
+        state: &mut BuildFragmentGraphState,
+    ) -> SchedulerResult<PbStreamNode> {
+        use risingwave_pb::stream_plan::*;
+
+        let stream_key = self
+            .stream_key()
+            .unwrap_or_else(|| {
+                panic!(
+                    "should always have a stream key in the stream plan but not, sub plan: {}",
+                    PlanRef::from(self.clone()).explain_to_string()
+                )
+            })
+            .iter()
+            .map(|x| *x as u32)
+            .collect_vec();
+
+        let source_catalog = self.source_catalog();
+        let backfill = SourceBackfillNode {
+            source_id: source_catalog.id,
+            source_name: source_catalog.name.clone(),
+            state_table: Some(
+                Self::infer_internal_table_catalog()
+                    .with_id(state.gen_table_id_wrapped())
+                    .to_internal_table_prost(),
+            ),
+            info: Some(source_catalog.info.clone()),
+            // XXX: what's the usage of this?
```

Comment:
You mean `row_id`? By design, rows from source should be assigned unique row_ids as its PK.